core/blockchain.go

```diff
@@ -1213,13 +1214,20 @@ func (bc *BlockChain) InsertReceiptChain(blockChain types.Blocks, receiptChain [
 				if frozen, _ := bc.db.Ancients(); frozen == 0 {
 					h := rawdb.ReadCanonicalHash(bc.db, 0)
 					b := rawdb.ReadBlock(bc.db, h, 0)
-					size += rawdb.WriteAncientBlock(bc.db, b, rawdb.ReadReceipts(bc.db, h, 0, bc.chainConfig), rawdb.ReadTd(bc.db, h, 0))
+					size += rawdb.WriteAncientBlock(aBatch, b, rawdb.ReadReceipts(bc.db, h, 0, bc.chainConfig), rawdb.ReadTd(bc.db, h, 0))
 					log.Info("Wrote genesis to ancients")
 				}
 			}
-			// Flush data into ancient database.
-			size += rawdb.WriteAncientBlock(bc.db, block, receiptChain[i], bc.GetTd(block.Hash(), block.NumberU64()))
+			// Cache the ancient data into batch
+			size += rawdb.WriteAncientBlock(aBatch, block, receiptChain[i], bc.GetTd(block.Hash(), block.NumberU64()))
 
+			// Flush the ancient batch if the cached data reachs the threshold
+			if aBatch.ValueSize() > ethdb.IdealAncientBatchSize {
+				if err := aBatch.Write(); err != nil {
+					return i, err
+				}
+				aBatch.Reset()
+			}
```

Comment:
It would be really nice if we can somehow avoid having the situation where e.g. the size of the `receipts` trigger flushes every second 'block', which flushes even the tiny things like `td`. If we can allow some portions to temporarily on-disk stray ahead of other parts, then we save a lot of syscalls. 
That's what we were discussing yesterday on discord.

Reply:
It's a nice idea. I have to think about how to change the code.